devito/ir/stree/algorithms.py

```diff
@@ -182,7 +189,33 @@ def preprocess(clusters, options=None, **kwargs):
                 processed.append(c.rebuild(exprs=[], ispace=ispace, syncs=syncs))
 
             halo_scheme = HaloScheme.union([c1.halo_scheme for c1 in found])
-            processed.append(c.rebuild(halo_scheme=halo_scheme))
+            itdims = c.ispace.itdims
+            if halo_scheme:
+                # We make sure that the halo scheme placement is valid for this cluster
+                # Mainly if a cluster has iteration dimension {t, d, f}
+                # and the halo_scheme has distributed_aindices {d} and
+                # loc_dirs {f} then the haloscheme needs to be lifted into its own
+                # {t, f} cluster because it needs to be inside an {f} loop while it would
+                # be placed outside the {d} loop and therefore outside the {f} loop if
+                # attached to the cluster.
+                dindex = max([itdims.index(d) for d in halo_scheme.loc_dirs
+                              if d in itdims] + [0])
+                aindex = max([itdims.index(d) for d in halo_scheme.distributed_aindices
+                              if d in itdims] + [0])
+                if dindex > aindex:
+                    hispace = c.ispace.project(itdims[:dindex])
+                else:
+                    hispace = None
+            else:
+                hispace = None
+
+            if hispace and options['mpi']:
```

Comment:
whatever happens here, and regardless of how we end up doing it, we should always do it, irrespective of whether MPI is enabled or not. This will confer robustness to our compiler and, in particular, it will avoid "bugs to only pop up with MPI enabled"